app/static/playlist.js

```diff
@@ -105,54 +142,107 @@ export default class PlaylistLabelRenderer {
-
-          if (element.label.work.is_context_indigenous) {
-            document.getElementById("indigenous").className =
-              "indigenous indigenous_active";
-          } else {
-            document.getElementById("indigenous").className = "indigenous";
-          }
-
-          if (labels.length > 1) {
-            // Update up next label
-            const elementNext = labels.find((label) => {
-              return label.label.id === this.state.nextLabelId;
-            });
-            document.getElementById("next_title").innerHTML =
-              elementNext.label.title;
-          }
-        }
-      }
+      location.hash = messageJson.label_id;
+    }
+  }
+
+  jumpToLabel(label_id) {
+    console.log('jump to label', label_id);
+    // Update the current state
+    this.state.currentLabelId = label_id;
+    const items = this.state.items;
+
+    // make a list of items that starts with the new label
+    const upcoming_items = [];
+    const start_index = items.findIndex(function(item) {
+      return item.label && item.label.id === label_id;
+    });
+    for (let i=0; i<items.length; i++) {
+      upcoming_items.push(items[(start_index + i) % items.length]);
+    };
+
+    this.state.upcomingItems = upcoming_items;
+
+    // update label content
+    this.updateMainLabelContent(upcoming_items[0]);
+    if (items.length > 1) {
+      this.state.nextLabelId = upcoming_items[1].label.id;
+      this.updateUpNextContent(upcoming_items);
+    }
+  }
+
+  updateMainLabelContent(item) {
+    const label = item.label;
+    // Update the label fields with the currently playing data
+    const title = document.getElementById("title").innerHTML = label.title;
+    this.addTitleAnnotation(label.work);
+    document.getElementById("subtitles").innerHTML = label.subtitles;
+    document.getElementById("content0").innerHTML = label.columns[0].content;
+    document.getElementById("content1").innerHTML = label.columns[1].content;
+    document.getElementById("content2").innerHTML = label.columns[2].content;
+
+    if (label.work.is_context_indigenous) {
+      document.getElementById("indigenous").className = "indigenous indigenous_active";
+    } else {
+      document.getElementById("indigenous").className = "indigenous";
+    }
+  }
+
+  updateUpNextContent(items) {
+    // Update up next label
+    const item = items[1];
+    document.querySelector("#next_title").innerHTML = item.label.title;
+
+    for (let i=1; i<items.length; i++) {
+      const label = items[i].label;
+      const id = "#up_next_label_" + i;
+      try {
+        document.querySelector(id + " .title").innerHTML = i + '. ' + label.title;
+        document.querySelector(id + " .subtitles").innerHTML = label.subtitles;
+      } catch(err) {}
+    }
+  }
+
+  updateProgress() {
+    const progressBar = document.getElementById("progress-bar");
+    const playbackPosition = this.state.playbackPosition;
+    progressBar.style.width = `${playbackPosition * 100}%`;
```

Comment:
Do you want to pull Harry's progress bar code in here to make it smoother? https://github.com/ACMILabs/playlist-label/pull/17/files#diff-9c4ac97c58888f532e32895a3c9de5f6R156

Reply:
There's a different ticket for that; I'll leave it for now.